runtime/src/main/cpp/CppSupport.hpp

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#ifndef RUNTIME_CPP_SUPPORT_H
+#define RUNTIME_CPP_SUPPORT_H
+
+#include <type_traits>
+#include <memory>
+
+// A collection of backported utilities from future C++ versions.
+
+namespace kotlin {
+
+////////////////////////// C++14 //////////////////////////
+
+template <typename T, typename... Args>
+std::unique_ptr<T> make_unique(Args... args) {
+    return std::unique_ptr<T>(new T(args...));
+}
+
+template <typename T>
+using make_unsigned_t = typename std::make_unsigned<T>::type;
+
+////////////////////////// C++17 //////////////////////////
+
+template <typename T>
+constexpr bool is_trivially_destructible_v = std::is_trivially_destructible<T>::value;
+template <typename T>
+constexpr bool is_nothrow_default_constructible_v = std::is_nothrow_default_constructible<T>::value;
+template <typename T>
+constexpr bool is_nothrow_destructible_v = std::is_nothrow_destructible<T>::value;
+template <typename T>
+constexpr bool is_copy_constructible_v = std::is_copy_constructible<T>::value;
+template <typename T>
+constexpr bool is_copy_assignable_v = std::is_copy_assignable<T>::value;
+template <typename T>
+constexpr bool is_move_constructible_v = std::is_move_constructible<T>::value;
+template <typename T>
+constexpr bool is_move_assignable_v = std::is_move_assignable<T>::value;
+template <typename T>
+constexpr bool is_nothrow_move_constructible_v = std::is_nothrow_move_constructible<T>::value;
+template <typename T>
+constexpr bool is_nothrow_move_assignable_v = std::is_nothrow_move_assignable<T>::value;
+
+} // namespace kotlin
+
```

Comment:
With some `#if __cplusplus < 201703L` and `using ...` we may switch between `std::` and `kotlin::std::` in the header depending on c++ version available for the current target. The point is: let the single code base to use C++17 `std` where available, and custom backport otherwise.

Reply:
Unfortunately that won't work: `__cplusplus` is defined by a compiler, not a standard library, and we're currently in a situation where our compiler is doing C++14, but not all stdlibs have full C++14.